1. Data (5)
1.1 Read Data (3)
As a first step, please load the datapoints from the given cluster_data.txt file using the inbuilt open() function. Note

In [1]:
def read_data(filepath):
    # TODO: read the file and store the datapoints in a list
    data = []
    with open(filepath, 'r') as f:
        for line in f:
            line = line.strip()
            line = line.split(",")
            data.append([float(coord) for coord in line])
    return data

1.2 Print Data (1)
Please read the data and print the first 5 elements of it.

In [2]:
filepath = "cluster_data.txt"
my_data = read_data(filepath)
for elem in my_data[:5]:
    print(elem)

[-2.672217304042017, -4.101281486217399]
[-8.039282818227496, 2.8319065022588026]
[3.3645318481024713, 0.31055081545006236]
[0.5534653004366121, 1.8002979493400275]
[-5.740691049309148, 3.9577428483393575]


1.3 Visualize Data (1)
For visualization we use the canonical matplotlib package. Please familiarize yourself with this package, even though we provide you with a visualization
function for now. We will for sure need it in a later stage.
The provided function accepts the data as input and optio

In [3]:
import numpy as np
print(np.ones((3,3)))

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [4]:
import numpy as np
import matplotlib.pyplot as plt

colors = ["g", "y", "c", "m", "b", "r", "k"]
def visualize_data(data, centroids=None, labels=None, figsize=(5, 4)):
    plt.figure(figsize=figsize)
    if labels is not None:
        color_data = [colors[x] for x in labels]
        if centroids is not None:
            color_centroid = colors[:len(centroids)]
    else:
        color_data = color_centroid = "C0"
    # plot the data
    plt.scatter([x[0] for x in data], [x[1] for x in data], c=color_data, s=20)
    # plot the centroids if they are given
    if centroids is not None:
        plt.scatter(
        [x[0] for x in centroids],
        [x[1] for x in centroids],
        c=color_centroid, marker="X",
        s=80, linewidths=1, edgecolors="k"
        )
        
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.show()

ModuleNotFoundError: No module named 'matplotlib'

2. Distance Function (2)
As we have learned in the lecture, we need some kind of metric to represent (dis-)similarity between data points. For now we will use the Euclidean distance,
which is defined for two points p and q with dimension n as
Euclidean distance formula
Please follow the function signature from below and complete the syntax, so that the function returns the Euclidean distance of two points. For now it is
sufficient if you implement it for points x ∈ R2.

3. K-means Algorithm (6)
The general procedure of the K-means algorithm is as follows:
1. You initialize the algorithm with k different centroids. The hyper-parameter k represents the number of clusters you assume in your dataset.
2. For each data point you assign the cluster with the nearest representative/centroid.
3. Based on the previously computed cluster assignments, you compute new centroids.
You repeat from step 2 until the centroids do not change anymore and the algorithm has converged.
Below you find a class KMeans that is initialized with a list of centroids. It implements two functions, assign_clusters and compute_new_centroids ,
which represent step 2. and 3. from above. Please implement the missing parts of this class. Only use pure python and the euclidean_distance function
from above.

4. Learning from Data (7)
4.1 Training loop (4)
Now we have everything together for setting up the training loop. Please perform k-means clustering for 4 epochs on the data with the three given centroids.
After each step (cluster assignment and new centroid computation), visualize the data, including labels and centroids.

4.2 Results (1)
What do you observe here? Do you see any problems? What are these problems and what do you need to change in order to get a better clustering of your
data?
Answer: We choose k = 3 even though the dataset has 4 clusters in total. This is why our algorithm converges to a sub-optimal solution. We would need to
initialize our algorithm with four centroids.
4.3 Hyper-parameter tuning (2)
In machine learning we have the so-called hyper-parameters. These are parameters we need to tune in order to get good results. For k-means we have a
single hyper-parameter k that defines how many centroids we use to initialize our algorithm. Below you find a function that returns randomly initialized
centroids, given the parameter k. Under the hood it uses a very important package called numpy, which we will further get to know in the upcoming lectures.
Feel free to already get familiar with it.